File: week1/pset1/credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>

int number_of_digit(long int n);
void Card_Type(long int n);
int Luhn_Algorithm(int count,long int n);
int calculate_sum1(int count,long int n);
int calculate_sum2(int count,long int n);

int main(void)
{
    long int n;
    int count,status;
    do
    {
        n = get_long("Number: ");
    }while(n<0);

    count = number_of_digit(n);

    if(count>=13 && count <=16)
    {
        status = Luhn_Algorithm(count,n);
        if(status == 1) Card_Type(n);
        else printf("INVALID\n");
    }
    else printf("INVALID\n");
}


int number_of_digit(long int n)
{
    int count=0;
    do
    {
        n/=10;
        ++count;
    }while(n!=0);
    return count;
}


int Luhn_Algorithm(int count,long int n)
{
    int sum1,sum2;

    sum1 = calculate_sum1(count,n);
    sum2 = calculate_sum2(count,n);
    printf("%d",sum1+sum2);
    if( (sum1+sum2)%10 == 0 ) return 1;

    else return 0;
}


int calculate_sum1(int count,long int n)
{
    int i,sum=0;
    long m=100,x=10,r;
    for(i=0;i<count/2;i++)
    {
        r=n%m;
        n=n-r;
        r = r/x;
        x*=100;
        m*=100;
        if(r*2>=10) sum = sum + r*2 - 9;
        else sum = sum + r*2;
    }
    return sum;
}


int calculate_sum2(int count,long int n)
{
    int i,sum=0;
    long m=1,x=10,r;

    if(count%2 != 0) count+=1;

    for(i=0;i<count/2;i++)
    {
        r=n%x;
        n=n-r;
        r = r/m;
        x*=100;
        m*=100;
        sum +=r;
    }
    return sum;

}


void Card_Type(long int n)
{
      //-> the VISA card have 13 or 16 digit and all Visa numbers start with 4
      // so i divided input number by 13 digit number to check the first number
      if(n/1000000000000 == 4 || n/1000000000000000 == 4) printf("VISA\n");

      //-> American Express have 15 digit and always start with 34 or 37
      // so i divided input number by a 14 digit number to check the first two numbers
      else if(n/10000000000000 == 34 || n/10000000000000 == 37) printf("AMEX\n");

      //-> MasterCard have 16 digit and always start with 51, 52, 53, 54, or 55
      // so i divided input number by a 15 digit number to check the first two numbers
      else if(n/100000000000000 == 51 || n/100000000000000 == 52
      || n/100000000000000 == 53 || n/100000000000000 == 54
      || n/100000000000000 == 55 || n/100000000000000 == 22)
      printf("MASTERCARD\n");

      else printf("INVALID\n");

}

File: week6/pset6/sentimental-credit/credit.py
from sys import exit

def main():
    number = str(get_value("Number: "))
    length = len(number)
    if length >= 13 and length <= 16:
        if not Luhn_Algorithm(number,length):
            print("INVALID")
            exit(1)

        if number[0] == "4":
            print("VISA")
            exit(0)

        elif number[0] == "3" and (number[1] == "4" or number[1] == "7"):
            print("AMEX")
            exit(0)

        elif number[0] == "5" and number[1] > "0" and number[1] < "6":
            print("MASTERCARD")
            exit(0)

        else:
            print("INVALID")
            exit(1)
            
    print("INVALID")
    exit(1)

def get_value(s):
    while True:
        try:
            n = int(input(s))
            if n > 0:
                return n
        except ValueError:
            continue

def Luhn_Algorithm(num, len):
    sum = 0

    if len%2 == 0:
        start1 = 0
        start2 = 1
    else:
        start1 = 1
        start2 = 0

    for i in range(start1, len, 2):
        tmp = int(num[i])*2
        if tmp >= 10:
            sum += tmp//10
            sum += tmp%10
        else:
            sum += tmp

    for i in range(start2 , len, 2):
        sum += int(num[i])

    if sum%10 == 0:
        return True

    return False

main()